Generate readfirstlane efficiently for uniform case

xuechen417 opened the pull request:
The Vulkan Enviroment for SPIR-V (section "Validation Rules within a
Module",  version 1.2.140) says that:
"If an instruction loads from or stores to a resource (including atomics
and image instructions)
and the resource descriptor being accessed is not dynamically uniform,
then the operand
corresponding to that resource (e.g. the pointer or sampled image
operand) must be decorated
with NonUniform."
The usage of NonUniform decoration is modified according to spec both in
CTS and glslang.
Example:
layout(set = 0, binding = 0) uniform texture2D uTex[];
layout(set = 1, binding = 0) uniform sampler uSamp;
layout(location = 0) flat in int Index;
layout(location = 0) out vec4 FragColor;
void main()
{
    // Old usage not according spec
    FragColor = texture(sampler2D(uTex[nonuniformEXT(Index)], Samp), vec2(0.5));
   // New usage according spec
   FragColor = texture(
   nonuniformEXT(sampler2D(uTex[Index], uSamp)), ec2(0.5));
}

it produces the following SPIR-V (only extract provided)
// Old usage
...
OpDecorate %18 NonUniform
OpDecorate %21 NonUniform
...
%5 = OpLabel
%17 = OpLoad %int %Index
%18 = OpCopyObject %int %17 <-- The index is NonUniform
%20 = OpAccessChain %_ptr_UniformConstant_10 %uTex %18
%21 = OpLoad %10 %20             <-- The image itself is NonUniform
%25 = OpLoad %22 %uSamp
%27 = OpSampledImage %26 %21 %25
%31 = OpImageSampleImplicitLod %v4float %27 %30 <-- The sampled image is *not* NonUniform
...
// New usage
...
OpDecorate %33 NonUniform
...
%10 = OpTypeImage %6 2D 0 0 0 1 Unknown
...
%31 = OpTypeSampledImage %10
...
%32 = OpSampledImage %31 %20 %30
%33 = OpCopyObject %31 %32
%37 = OpImageSampleImplicitLod %7 %33 %36
...

Fixes: 
- https://github.com/GPUOpen-Drivers/llpc/issues/840
- dEQP-VK.descriptor_indexing.sampled_image*.

The reason of cts failutres is that the descriptor index is mistook as uniform in indexDescPtr which it should be non-uniform.
Solution:
- Remove the non-uniform flag in indexDescPtr, which can make the test pass.
- Generate readfirstlane for operands without non-uniform decoration. 
  1) Insert readerfirstlane on the resource descriptor initially
  2) Add a otpimization pass in middle-end to push up readfirstlane to transform the use of vgpr to sgpr  as many as possible.

xuechen417 commented:
@nhaehnle This new PR changes the way to lift readfirstlane by processing per-block for each instruction in a reverse order not in a whole module, which is suggested by @ruiling. The original PR #811 is closed. Please reivew it.

amdvlk-admin commented:
### Test summary for commit 3097bf78451a2b2073a69f8f082aacb1e3a81518
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 20bc469e9806bb60b22559f727c1321354091898</li>
<li>SPVGEN: dev c054813a9a894e32aaaa04c6717a667c15f60cfd</li>
<li>XGL: dev feb87bf013f2b4bd16f7a1da39a5a97fdaa7470d</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev d3f03b36d914d6f4c4d27709bf29b89c4803c9cf</li>
</ul></details>

<details><summary>CTS tests (Failed: 2/131664)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22981/43895 (52.4%)</li>
<li>Failed: 0/43895 (0.0%)</li>
<li>Skipped: 20914/43895 (47.6%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22979/43895 (52.3%)</li>
<li><details><summary>Failed: 2/43895 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.pipeline_barrier.host_write_storage_texel_buffer.8192
Stack trace: Flake

FAILURE: dEQP-VK.synchronization.win32_keyed_mutex.write_image_fragment_read_blit_image.image_128x128_r16_uint_nt
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 20914/43895 (47.6%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21622/43874 (49.3%)</li>
<li>Failed: 0/43874 (0.0%)</li>
<li>Skipped: 22252/43874 (50.7%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 73338c170f18aec903d6150184375e707b1ba6e5
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 20bc469e9806bb60b22559f727c1321354091898</li>
<li>SPVGEN: dev c054813a9a894e32aaaa04c6717a667c15f60cfd</li>
<li>XGL: dev feb87bf013f2b4bd16f7a1da39a5a97fdaa7470d</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev d3f03b36d914d6f4c4d27709bf29b89c4803c9cf</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/131664)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22981/43895 (52.4%)</li>
<li>Failed: 0/43895 (0.0%)</li>
<li>Skipped: 20914/43895 (47.6%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22980/43895 (52.4%)</li>
<li>Failed: 0/43895 (0.0%)</li>
<li>Skipped: 20915/43895 (47.6%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21622/43874 (49.3%)</li>
<li>Failed: 0/43874 (0.0%)</li>
<li>Skipped: 22252/43874 (50.7%)</li>
</ul></details>


</ul></details>

ruiling commented:
Let me try to explain my idea. The problem I am trying to solve is how to efficiently checking whether all users want to be treated as scalar. I think a per-block algorithm (only checking against users in the same block) seems enough for most cases. So a per-block solution is a good first-step, we can extend it to handle cross-block version later if needed. A per-block solution has a very good attribute that we can easily prove the algorithm is correct. By visiting instructions from back to front within a block, we can always be sure that all users will be visited and checked before its define instruction. Consider a simple example (suppose we plan to move all the instructions into scalar from):
Ins1: %1 = ...
Ins2: %2 = ... %1
Ins3: %3 = ... %1
Ins4: %4 = ... %2
Ins5: %5 = ... %2
Ins6: readfirstlane %3
Ins7: readfirstlane %4
Ins8: readfirstlane %5
The existing solution cannot make sure the instructions can be visited in order 5/4 before 3/2 and then 1. We should not query whether all users of %1 want to be scalar before Ins2&Ins3 visited because we would get in-accurate result.
And also for per-block solution, we only need to check for uniformity against the set of scalar candidates in the same block, we don't need to check against all the candidates in the whole function.

ruiling commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+        for (unsigned idx = 0; idx < numOps; ++idx) {
+          Value *operand = candidateInst->getOperand(idx);
+          if (auto operandInst = dyn_cast<Instruction>(operand)) {
+            // A readfirstlane lift should occur in one block
+            // Stop propagtion if one of the operands isn't in the same block
+            if (operandInst->getParent() != curBb) {
+              liftInsts.insert(candidateInst);
+              break;
+            }
+
+            // Skip the marked instuction
+            if (markedUniforms.find(operandInst) != markedUniforms.end())
+              continue;
+
+            // Check the operand that cannot be deduced as uniform by divergence analysis
+            if (!m_divergenceAnalysis->isUniform(operandInst)) {
+              // The qualified intruction must has all users marked as uniform
+              bool allUsersUniform = true;
+              for (auto user : operandInst->users()) {
+                auto userInst = dyn_cast<Instruction>(user);
+                if (markedUniforms.find(userInst) == markedUniforms.end()) {
+                  allUsersUniform = false;
+                  break;
+                }
+              }
+              if (allUsersUniform) {
+                auto loadInst = dyn_cast<LoadInst>(operandInst);
+                if (dyn_cast<CallInst>(operandInst) || (loadInst && loadInst->getType()->isIntegerTy(32))) {
+                  // Stop propagation
+                  liftInsts.insert(operandInst);
+                } else {
+                  // Mark the operand as uniform
+                  markedUniforms.insert(operandInst);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
```
I have made some modification to this for-loop. the main change is query allUsersMarkedUniform on the def of the current instruction instead of query the operand. because query on the operand may get in-complete result. @xuechen417 @nhaehnle could you help take a look whether this basically looks good? which way do you prefer?
please note that I didn't compile the code, so there may be some build error.
```
/// check whether all operands of \p inst defined in \p block
static bool isAllOperandsDefinedInBlock(Instruction *inst, BasicBlock *block) {
  unsigned numOps = inst->getNumOperands();
  for (unsigned idx = 0; idx < numOps; ++idx) {
    Value *operand = inst->getOperand(idx);
    if (auto operandInst = dyn_cast<Instruction>(operand)) {
      if (operandInst->getParent() != block)
        return false;
    }
  }
  return true;
}
/// Check whether all users of \p inst has been marked as uniform (i.e. been pushed in \p markedUniforms)
static bool isAllUsersMarkedUniform(Instruction *inst, DenseSet<Instruction *> &markedUniforms) {
  for (auto user : inst->users()) {
    auto userInst = dyn_cast<Instruction>(user);
    if (markedUniforms.find(userInst) == markedUniforms.end())
      return false;
  }
  return true;
}
/// a simple heuristic for whether it is beneficial to stop propagate uniform property
bool PatchReadFirstLane::isBeneficialToStop(Instruction *inst) {
  unsigned numOps = inst->getNumOperands();
  if (numOps > 2)
    return true;
  if (numOps == 2 && m_divergenceAnalysis->isDivergent(inst->getOperand(0)) &&
      m_divergenceAnalysis->isDivergent(inst->getOperand(1)))
    return true;
  return false;
}


   BasicBlock::reverse_iterator instIt = curBb->rbegin(), endIt = curBb->rend();
   // skip the instructions we don't care
   for (; instIt != endIt; ++instIt)
      if (MarkedUniforms.find(&*instIt) != MarkedUniforms.end())
        break;

   for (; instIt != endIt; ++instIt) {
      Instruction *candidate = &*instIt;
      if (!m_divergenceAnalysis->isUniform(candidate)) {
        // The candidate instruction is non-uniform, check whether all users of it
        // want to be promoted to uniform
        if (isAllUsersMarkedUniform(candidate, markedUniforms)) {
          auto loadInst = dyn_cast<LoadInst>(candidate);
          // why do we need to stop at i32 type???
          if (isa<CallInst>(candidate) || (loadInst && loadInst->getType()->isIntegerTy(32))) {
            // Stop propagation when the instruction has unknown side-effect
            liftInsts.insert(candidate);
          } else if (isBeneficialToStop(candidate)) {
            // Stop propagation when it is not worth lifting a readfirstlane
            liftInsts.insert(candidate);
          } else if (!isAllOperandsDefinedInBlock(candidate, curBb))
            // stop propagation if some operand is defined in other blocks
            liftInsts.insert(candidate);
          else
            markedUniforms.insert(candidate);
        } else {
          // TODO: check if we can/need insert readfirstlane before the user later.
          for (auto user : candidate->users()) {
            auto userInst = dyn_cast<Instruction>(user);
            if (userInst->getParent() != curBb)
              continue;
            if (markedUniforms.find(userInst) != markedUniforms.end()) {
              liftInsts.insert(userInst);
            }
          }
        }
      }
    }
```

nhaehnle replied:
The helper functions mostly make sense to me, but I'd tweak the isBeneficialToStop heuristic slightly. I would say: call it "beneficial to stop" if there is more than one non-uniform operand. The reason for this is that you can have a `getelementptr` with many operands. As long as only one of them is divergent, it makes sense to lift the readfirstlane.

You need to stop propagation when you encounter an instruction for which `isSourceOfDivergence` is true: lifting a readfirstlane above such an instruction will give an incorrect result.

I'm still trying to wrap my head around the overall logic of the loop though.

What benefit is there from iterating separately over the basic blocks? Iterating over the instructions of a basic block would make sense if that is your approach for finding `readfirstlane` instructions. However, since you find those instructions via the module-based scan, we could then just rely on a "work queue" approach to look at only the instructions that the algorithm really needs to inspect.

That is, when the algorithm finds a new instruction which is a potential candidate for a readfirstlane being lifted over it, add that instruction to a FIFO. Initialize the FIFO with the initial readfirstlanes. Using a FIFO should ensure that you only need to look at each relevant instruction once. 

Also, I believe the inner loop should be structured as a loop over operands:

```
  Use *divergentUse = nullptr;
  for (Use *use : candidateInst->uses()) {
    if (!m_divergenceAnalysis->isDivergentUse(use))
      continue;
    if (divergentUse) {
      have two divergent uses here: do not lift readfirstlane over candidateInst
    }
    divergentUse = use;
  }
  if (Instruction *divergentDef = dyn_cast<Instruction*>(divergentUse->getValue())) {
    if (divergentDef->getParent() == candidateInst->getParent()) {
      add divergentDef as a new candidate instruction
      continue; // go to the next candidate instruction
    }
  }
  we can still lift the readfirstlane above candidateInst, but we have to make it
  a readfirstlane in the same basic block, and we don't add divergentDef to be
  a new candidate instruction 
  a special case if when divergentUse->getValue() is a function argument; in that case,
  we can put the readfirstlane into the function entry block
```
Does that help? The details will need a bit more thought.

xuechen417 replied:
Yes, it helps a lot. I'm re-working on this part code as your suggestion.
I have one question that image resource type workaround for gfx10 gives a special case:
```
%.i3 = extractelement <8 x i32> %33, i32 3
...
%39 = icmp sgt i32 %.i3, -1
%40 = and i32 %.i3, 268435455
%41 = select i1 %39, i32 %40, i32 %.i3
%42 = call i32 @llvm.amdgcn.readfirstlane(i32 %.i0)
%43 = call i32 @llvm.amdgcn.readfirstlane(i32 %.i11)
%44 = call i32 @llvm.amdgcn.readfirstlane(i32 %.i2)
%45 = call i32 @llvm.amdgcn.readfirstlane(i32 %41)
```
If we use the stop condition "there is more than one non-uniform operand". We cannot do a lifting.
I discussed with @ruiling and we think this case should be handled. For more than one non-uniform operand, if they can be deduced to one non-uniform, we can lift the readfirstlane over it. What do you think?

ruiling replied:
> The helper functions mostly make sense to me, but I'd tweak the isBeneficialToStop heuristic slightly. I would say: call it "beneficial to stop" if there is more than one non-uniform operand. The reason for this is that you can have a `getelementptr` with many operands. As long as only one of them is divergent, it makes sense to lift the readfirstlane.
> 
The beneficial() function may need further improved, if we prove that we are not inserting more readfirstlane, we are ok to go lift the readfirstlane.

> You need to stop propagation when you encounter an instruction for which `isSourceOfDivergence` is true: lifting a readfirstlane above such an instruction will give an incorrect result.
Yes, that's true.

> 
> I'm still trying to wrap my head around the overall logic of the loop though.
> 
> What benefit is there from iterating separately over the basic blocks? Iterating over the instructions of a basic block would make sense if that is your approach for finding `readfirstlane` instructions. 
You can see my previous explaination under this PR, The reason is try to make sure we have already make sure the users are already visited so we can get correct query result whether all users want to be lifted. 
```
Ins1: %1 = ...
Ins2: %2 = ... %1
Ins3: %3 = ... %1
Ins4: %4 = ... %2
Ins5: %5 = ... %2
Ins6: readfirstlane %3
Ins7: readfirstlane %4
Ins8: readfirstlane %5
```
In the above example, if we want to know whether all users of `%2` want to be optimized as uniform, we need to make sure we have already visited all users of `%2`. So a backward traversal seems a good fit for this, although I admit this would introduce visiting unnecessary instructions :-( Later I realized the current implementation can still get correct query result at the most-closest user because other users already visited at this point.

> However, since you find those instructions via the module-based scan, we could then just rely on a "work queue" approach to look at only the instructions that the algorithm really needs to inspect.

> 
> That is, when the algorithm finds a new instruction which is a potential candidate for a readfirstlane being lifted over it, add that instruction to a FIFO. Initialize the FIFO with the initial readfirstlanes. Using a FIFO should ensure that you only need to look at each relevant instruction once.

> 
> Also, I believe the inner loop should be structured as a loop over operands:
So below part of code are all contained within the loop over operands?
```
for (unsigned idx = 0; idx < numOps; ++idx) {
```
If this is true, may be `candidateInst` means `operandInst`?
> 
> ```
>   Use *divergentUse = nullptr;
>   for (Use *use : candidateInst->uses()) {
>     if (!m_divergenceAnalysis->isDivergentUse(use))
>       continue;
Why using `isDivergentUse()` instead of find in `markedUniform` here? divergentUse means that either the value (candidateInst) is divergent, or it is temporary divergent at the use place. I am not sure what do you want to achieve here?

>     if (divergentUse) {
>       have two divergent uses here: do not lift readfirstlane over candidateInst
>     }
>     divergentUse = use;
>   }
>   if (Instruction *divergentDef = dyn_cast<Instruction*>(divergentUse->getValue())) {
>     if (divergentDef->getParent() == candidateInst->getParent()) {
>       add divergentDef as a new candidate instruction
>       continue; // go to the next candidate instruction
>     }
>   }
>   we can still lift the readfirstlane above candidateInst, but we have to make it
>   a readfirstlane in the same basic block, and we don't add divergentDef to be
>   a new candidate instruction 
>   note that this also covers the special case where divergentUse->getValue()
>   is a function argument
> ```
> 
> Does that help? The details will need a bit more thought.
> 
> Edit: I initially wrote that you can put readfirstlanes into the function entry block if the divergent use value is a function argument. This was incorrect, the readfirstlane must always be in the same basic block as the user.

nhaehnle replied:
> > What benefit is there from iterating separately over the basic blocks? Iterating over the instructions of a basic block would make sense if that is your approach for finding `readfirstlane` instructions.
> 
> You can see my previous explaination under this PR, The reason is try to make sure we have already make sure the users are already visited so we can get correct query result whether all users want to be lifted.
> 
> ```
> Ins1: %1 = ...
> Ins2: %2 = ... %1
> Ins3: %3 = ... %1
> Ins4: %4 = ... %2
> Ins5: %5 = ... %2
> Ins6: readfirstlane %3
> Ins7: readfirstlane %4
> Ins8: readfirstlane %5
> ```
> 
> In the above example, if we want to know whether all users of `%2` want to be optimized as uniform, we need to make sure we have already visited all users of `%2`. So a backward traversal seems a good fit for this, although I admit this would introduce visiting unnecessary instructions :-( Later I realized the current implementation can still get correct query result at the most-closest user because other users already visited at this point.

I think you make a good point here, actually. I wanted to avoid visiting unnecessary instructions, and I was thinking that by putting instructions into a FIFO in the appropriate way we would be visiting them in the correct order, but I was wrong about that.

Walking the basic block backwards is a simple approach to guarantee correct visit order.

An alternative would be to use an `std::priority_queue` that uses `!Instruction::comesBefore` as its comparison operator. You could add candidates into the priority queue as they become available, and the queue would be able to always give you the last available candidate. The advantage is that it skips over irrelevant instructions. The disadvantage is the increased complexity.

I'd be fine with either solution.


> If this is true, may be `candidateInst` means `operandInst`?

Sorry, in the loop below I should have used `candidateInst->operands()`:

> > ```
> >   Use *divergentUse = nullptr;
> >   for (Use *use : candidateInst->operands()) {
> >     if (!m_divergenceAnalysis->isDivergentUse(use))
> >       continue;
> > ```
> 
> Why using `isDivergentUse()` instead of find in `markedUniform` here? divergentUse means that either the value (candidateInst) is divergent, or it is temporary divergent at the use place. I am not sure what do you want to achieve here?

This is meant to be part of the heuristic for deciding whether to lift readfirstlane above `candidateInst`. Example:
```
  %x = add i32 %a, %b    <---- candidateInst
  %y = call i32 @llvm.amdgcn.readfirstlane(i32 %x)
```
The loop iterates over the uses of `candidateInst`, which are uses of `%a` and of `%b`. If one of them is already known to be uniform, then we don't need to apply readfirstlane to it.

Though there's the problem of the descriptor with fixups. There are two possible ways I can see of addressing this:

1. Ask ourselves why the code comes out this way in the first place. Maybe earlier, when the readfirstlane is first inserted, we can already insert it before the fixup code instead of after it? That would allow you to avoid this case in PatchReadFirstLane.

2. Solve it directly. Do you have a proposal for how that could be done algorithmically?

The most general form of the solution is I believe to first propagate always, and then to figure out all the nodes in the data dependency DAG to which readfirstlane should be applied to minimize the number of readfirstlanes. This problem in turn is a type of minimum cut problem: the nodes to which readfirstlane can be applied form a DAG, and a minimum (node) cut gives you a minimum set of nodes to which to apply readfirstlane. Appropriate weights still need to be added. Min-Cut is a problem that can be solved reasonably fast, but good solvers are non-trivial pieces of code.

With that background, perhaps you're thinking of a convenient heuristic?

xuechen417 replied:
Yes, I tried a heuristic. It seems work.  To describe clearly, I call a divergent use of `candidateInst` as resultInst.
The available lift is each divergent use of a resultInst is either from resultInst array or from a fixed divergent.
The fixed divergent is from the first resultInst which should have only one divergent use after sorting by comeBefore.
I'm testing the new implementation and will update the PR soon.

ruiling commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+        Instruction *inst = cast<Instruction>(user);
+        if (inst->getFunction() != &function)
+          continue;
+        initialReadFirstLanes.insert(inst);
```
Putting the initialReadFirstLanes that may belong to different blocks seems not a good idea. I think a map<BasicBlock*, DenseSet<Instruction *>> may help here, you can put the instructions that belong to the same block together.

xuechen417 replied:
Got it. Done.

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+      // The candidate instrction for a lift should be marked as uniform
+      if (markedUniforms.find(candidateInst) != markedUniforms.end()) {
+        unsigned numOps = candidateInst->getNumOperands();
+        if (dyn_cast<CallInst>(candidateInst))
+          --numOps;
```
Is this really correct? I don't think the caller is counted among `getNumOperands`, but even if it does, the count has to be consistent for iterating over the operands, right?

xuechen417 replied:
- Yes, `getNumOperands` returns 2 for `call i32 @llvm.amdgcn.readfirstlane(i32 %.i35)`. 
- Yes, the count can be consistent because the return value has already marked as uniform.

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+              }
+              if (allUsersUniform) {
+                auto loadInst = dyn_cast<LoadInst>(operandInst);
+                if (dyn_cast<CallInst>(operandInst) || (loadInst && loadInst->getType()->isIntegerTy(32))) {
+                  // Stop propagation
+                  liftInsts.insert(operandInst);
```
I don't understand why you stop propagation here.

xuechen417 replied:
I think the index may from a function call or a loadInst. So I set the them as the stop condition.

xuechen417 commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+using namespace lgc;
+using namespace llvm;
+
+namespace llvm {
+class LegacyDivergenceAnalysis;
+} // namespace llvm
```
Got it. Removed

xuechen417 commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+//
+// @param [in,out] analysisUsage : The place to record our analysis pass usage requirements.
+void PatchReadFirstLane::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
+  analysisUsage.addRequired<LegacyDivergenceAnalysis>();
```
Got it. Done.

xuechen417 replied:
Fixed.

xuechen417 commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+}
+
+// =====================================================================================================================
+// Try to lift readfirstlanes per block to transform SGPR instructions into VGPR instruction.
```
Sorry, it's a typo. Fixed.

xuechen417 commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+  // Try to lift readfirstlanes per block
+  for (auto curBb : readFirstLaneBlocks) {
+    // Collect the instructions that can be marked as uniform
+    DenseSet<Instruction *> markedUniforms(initialReadFirstLanes);
```
Got it. Fixed.

xuechen417 commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,245 @@
+
+// =====================================================================================================================
+// Initializes the pass of LLVM patching operations for readfirstlane optimizations.
+INITIALIZE_PASS(PatchReadFirstLane, DEBUG_TYPE, "Patch LLVM for readfirstlane optimizations", false, false)
```
Got it.

xuechen417 commented on lgc/builder/ImageBuilder.cpp:
```diff
@@ -1830,3 +1851,22 @@ Value *ImageBuilder::handleFragCoordViewIndex(Value *coord, unsigned flags, unsi
+    elem = CreateIntrinsic(Intrinsic::amdgcn_readfirstlane, {}, elem);
+    newDesc = CreateInsertElement(newDesc, elem, elemIdx);
+  }
+  origDesc->replaceUsesWithIf(newDesc, [imageInst](Use &U) { return U.getUser() == imageInst; });
```
Yes,  I will try it.

xuechen417 reviewed (commented):
V2: re-implement liftReadFirstLane based on all comments.

amdvlk-admin commented:
### Test summary for commit 869ebf0bd86de214733aeb363d312f0c23aa3dda
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 85f858393412732198fcd3dc0a0efb8653f5d3f9</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 0d602bbcfa1a86fd52f66966554b05e384a10d38</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1598ada40008302abedfd724275329f8e7dea637</li>
</ul></details>

<details><summary>CTS tests (Failed: 3/140859)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 24005/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22954/46960 (48.9%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 24004/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22955/46960 (48.9%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 22646/46939 (48.2%)</li>
<li><details><summary>Failed: 1/46939 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 24292/46939 (51.8%)</li>
</ul></details>


</ul></details>

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,355 @@
+bool PatchReadFirstLane::applyReadFirstLane(Instruction *inst, BuilderBase &builder) {
+  // Here, we insert amdgcn_readfirstlane only if the type is float or int
+  // TODO: Support various types when backend work is ready
+  Type *uniformTy = inst->getType();
+  bool isFloatTy = uniformTy->isFloatTy();
+  if (!isFloatTy && !uniformTy->isIntegerTy(32))
+    return false;
+
```
It would be better to check this earlier. That is, applyReadFirstLane should just check the types via an assertion, and earlier code should make sure that applyReadFirstLane is never called with types for which it doesn't apply. This helps us avoid weird inefficiencies.

nhaehnle reviewed (commented):
Thanks for the new version. I think I understand where you're trying to go with the heuristic, and I like it, but I think the code could be reorganized to make the important ideas clearer and avoid the implicitly nested loops. As it is, there are still too many vague identifiers in the code, such as `takeAction`.

Trying to rephrase your key idea in my own words: if an instruction that can be assumed to be uniform has multiple divergent operands, then you take the definition of the divergent operand that is earliest in basic block order (call it "the earliest divergent operand") and propagate up to **that** instruction; if it turns out that that instruction can be assumed to be uniform, then we can just insert the `readfirstlane` there (or propagate).

In terms of what I wrote before about minimum cuts, your heuristic finds a minimum cut if it's of size 1. If it can't find one, it just gives up. This is a good trade-off.

With this in mind:

We say that a value `V` can be "assumed uniform" if the value itself isn't known to be uniform (by DivergenceAnalysis), but we can replace all of its uses by `readfirstlane(V)` and still be correct.

Actually doing this replacement is only worth it under certain conditions according to the heuristic.

So I think the code could be made clearer if the logic was split up into three steps in each basic block:

1. Collect all instructions that "can be assumed uniform" in a set (`m_canAssumeUniform`).
2. Determine the best places to insert readfirstlane according to your heuristic.
3. Apply readFirstLane on all determined locations.

Step 1 is just the propagation, which is done for all divergent operands (and as long as we stay in the same basic block). Correctness checks go here, e.g. the `isSourceOfDivergence` check.

Step 2 is where your heuristic comes in, but now you don't always have to re-check whether all the conditions are satisfied.

Does this make sense?

xuechen417 commented:
> Thanks for the new version. I think I understand where you're trying to go with the heuristic, and I like it, but I think the code could be reorganized to make the important ideas clearer and avoid the implicitly nested loops. As it is, there are still too many vague identifiers in the code, such as `takeAction`.
> 
> Trying to rephrase your key idea in my own words: if an instruction that can be assumed to be uniform has multiple divergent operands, then you take the definition of the divergent operand that is earliest in basic block order (call it "the earliest divergent operand") and propagate up to **that** instruction; if it turns out that that instruction can be assumed to be uniform, then we can just insert the `readfirstlane` there (or propagate).
> 
> In terms of what I wrote before about minimum cuts, your heuristic finds a minimum cut if it's of size 1. If it can't find one, it just gives up. This is a good trade-off.
> 
> With this in mind:
> 
> We say that a value `V` can be "assumed uniform" if the value itself isn't known to be uniform (by DivergenceAnalysis), but we can replace all of its uses by `readfirstlane(V)` and still be correct.
> 
> Actually doing this replacement is only worth it under certain conditions according to the heuristic.
> 
> So I think the code could be made clearer if the logic was split up into three steps in each basic block:
> 
> 1. Collect all instructions that "can be assumed uniform" in a set (`m_canAssumeUniform`).
> 2. Determine the best places to insert readfirstlane according to your heuristic.
> 3. Apply readFirstLane on all determined locations.
> 
> Step 1 is just the propagation, which is done for all divergent operands (and as long as we stay in the same basic block). Correctness checks go here, e.g. the `isSourceOfDivergence` check.
> 
> Step 2 is where your heuristic comes in, but now you don't always have to re-check whether all the conditions are satisfied.
> 
> Does this make sense?

Yes, it makes the logic clearer. 👍  I'm trying to reorganize the logic as your suggesed.  
Thanks a lot for giving an accurate summary of the heuristic.

xuechen417 reviewed (commented):
V3: Reorganize the logic in liftReadFirstLane with three steps; Update PatchInvalidImageDescriptor.lgc test.

xuechen417 commented:
V4: Format

amdvlk-admin commented:
### Test summary for commit 8ad4cc2bfb0544b99bbd84d39c3bbf181e3c4e89
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 007816f4bd2fd3ace30f8af9629cff7ec1dcb998</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev b25c1eac19f1773b753ed727314aa3cf2c0897dd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 5cb4dc49f6f35597873e6255129b92d102aae22b</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24701/46939 (52.6%)</li>
<li><details><summary>Failed: 1/46939 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.pipeline.image.suballocation.sampling_type.separate.view_type.cube_array.format.r32_sfloat.count_8.size.13x13_array_of_36
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,383 @@
+
+  LegacyDivergenceAnalysis *m_divergenceAnalysis; // The divergence analysis
+  TargetTransformInfo *m_targetTransformInfo;     // The target tranform info to determine stop propagation
+  DenseMap<Instruction *, DenseSet<Instruction *>>
+      m_canAssumeUniformDivergentUsesMap;    // The instructions that can be assumed uniform
```
I suggest making the value type a `SmallVector<Instruction *, 2>`: you don't *really* need it to be a set (even though there could perhaps be duplicates, but not typically), and making it a SmallVector is easier on memory allocations.

Also, can we improve the comment? Perhaps: "The instructions that can be assumed to be uniform. If the vector associated to an instruction I is non-empty, it contains the instructions that, if readfirstlane is applied to them, lead to I being uniform."

Is that accurate? Something like that should make the meaning of the variable clear and helps us reason about it.

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,383 @@
+  for (BasicBlock::reverse_iterator instIt = block->rbegin(), endIt = block->rend(); instIt != endIt; ++instIt) {
+    Instruction *inst = &*instIt;
+    // Skip readfirstlane-irrelevant instructions
+    if (canAssumeUniforms.count(inst) == 0)
+      continue;
```
Should the isSourceOfDivergence check go here instead of in `isBestInsertLocation`? The argument would be: If instruction I can be assumed uniform, but it is a source of divergence, then it's not clear that its operands can also automatically be assumed to be uniform.

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,383 @@
+      continue;
+    }
+    Instruction *earliestDivergentDef = *divergentDefSet.begin();
+    if (divergentDefCount == 1 && m_canAssumeUniformDivergentUsesMap.count(earliestDivergentDef) == 0) {
+      earliestDivergentDef = nullptr;
```
Can this special case for divergentDefCount == 1 somehow be folded into the general case?

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,383 @@
+      auto earliestUniformDivergentUsesIt = m_canAssumeUniformDivergentUsesMap.find(earliestDivergentDef);
+      bool canPropagateUp = (earliestUniformDivergentUsesIt != m_canAssumeUniformDivergentUsesMap.end() &&
+                             earliestUniformDivergentUsesIt->second.size() == 1);
+      if (canPropagateUp) {
+        DenseSet<Instruction *> divergentUses;
+        for (auto inst : divergentDefSet) {
+          auto assumeUniformOperandsIt = m_canAssumeUniformDivergentUsesMap.find(inst);
+          if (assumeUniformOperandsIt == m_canAssumeUniformDivergentUsesMap.end()) {
+            canPropagateUp = false;
+            break;
+          }
+          for (auto operand : assumeUniformOperandsIt->second) {
+            divergentUses.insert(operand);
+          }
+        }
+        if (canPropagateUp) {
+          Instruction *earliestOperand = *(earliestUniformDivergentUsesIt->second.begin());
+          divergentDefSet.insert(earliestOperand);
+          for (auto use : divergentUses) {
+            if (divergentDefSet.count(use) == 0) {
+              canPropagateUp = false;
+              break;
+            }
+          }
+        }
```
Doesn't this have to transitively check instructions somehow? Like in the example @ruiling gave in https://github.com/GPUOpen-Drivers/llpc/pull/976#issuecomment-699591447

nhaehnle reviewed (commented):
Thanks, I find this much clearer! I do have a few more algorithmic concerns though.

xuechen417 reviewed (commented):
V5: Address review comments from @nhaehnle

amdvlk-admin commented:
### Test summary for commit 8d9b37e3fd57209ad75148fef1c0d8e7e97cf7b7
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c9860f1b279efa64b57714a17025e32d12dbb593</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

xuechen417 commented:
V6: Formatting and update PatchInvalidImageDescriptor.lgc

amdvlk-admin commented:
### Test summary for commit 5c23e704d7cede28460644a66d2c55a2872efb4e
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c9860f1b279efa64b57714a17025e32d12dbb593</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

ruiling reviewed (commented):
> So I think the code could be made clearer if the logic was split up into three steps in each basic block:
> 
>     1. Collect all instructions that "can be assumed uniform" in a set (`m_canAssumeUniform`).
> 
>     2. Determine the best places to insert readfirstlane according to your heuristic.
> 
>     3. Apply readFirstLane on all determined locations.
> 
> 
> Step 1 is just the propagation, which is done for all divergent operands (and as long as we stay in the same basic block). Correctness checks go here, e.g. the `isSourceOfDivergence` check.
> 
> Step 2 is where your heuristic comes in, but now you don't always have to re-check whether all the conditions are satisfied.

I agree to split the algorithm into three steps if we can do as described above. Currently seems  there are still some correctness checking (I mean the fat-pointer stuff check) in the second step. I think correctness checking should all go to first step?

I remember we conservatively stop propagation for `CallInst` previously, seems removed in latest version. Is there any reason this got removed?

xuechen417 commented:
V7: Addressed comment from @ruiling . Put fat-pointer check to the first step. CallInst check is not explictly checked because it can be excluded by `isSourceDivegence`.

amdvlk-admin commented:
### Test summary for commit b72de4b91d9ca4675fb0316fc9d6dc059707d3fd
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c9860f1b279efa64b57714a17025e32d12dbb593</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

xuechen417 commented:
@nhaehnle The CTS failures are expected to be fixed at the end of November. Could you take a look? Thanks.

xuechen417 commented:
V8: Rebase. The CI failures with assertion on should be fixed by "Fixes to waterfall multi-begin support" 303a3fd.

amdvlk-admin commented:
### Test summary for commit 7dd021d778cf6504b12f3e38fa7adddc2f0692e7
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 16871e90b0f2e6937aecee415af4d2556384bd7f</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev f623fc7ca78c3cec2123a16700a8819f3ccf0656</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 6d7376cb11eeeb48833cd87a055249b7e0a8d9b5</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/144501)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26321/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21853/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26319/48174 (54.6%)</li>
<li><details><summary>Failed: 1/48174 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.rasterization.interpolation.projected.non_strict_lines
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 21854/48174 (45.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24852/48153 (51.6%)</li>
<li>Failed: 0/48153 (0.0%)</li>
<li>Skipped: 23301/48153 (48.4%)</li>
</ul></details>


</ul></details>

dstutt commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,418 @@
+}
+
+// =====================================================================================================================
+// Returns true if all users of the given instruction are already readfirtlane
```
Spelling

dstutt commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,418 @@
+  DenseSet<Instruction *> visitedCandidates;
+  for (auto &initialReadFirstLane : initialReadFirstLanes) {
+    // Find an insert location for each initial readfirstlane
+    std::deque<Instruction *> candidateInstQueuqe;
```
Name seems a bit strange: candidateInstDeque?

xuechen417 replied:
Fixed the typo.

dstutt reviewed (commented):
This LGTM - a couple of very very minor niggles.

I haven't fully understood what's actually being done here.
@nhaehnle - are you happy with the state of this change now? (It currently passes testing as a result of some more fixes in the waterfall intrinsic handling). If you haven't got time I could take another look to get a deeper understanding.

xuechen417 commented:
@dstutt Thanks for your review. I'm refining the code and adding more comments. I will push the update ASAP. Could you review after that?

ruiling commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,418 @@
+    // For all operands, see if all operands can be assumed uniform and are in the same block. If true, we can collect
+    // the operands to be used for the propagation heuristic in findInsertLocation.
+    SmallVector<Instruction *, 2> operandInsts;
+    for (Use &use : inst.operands()) {
+      if (auto operandInst = dyn_cast<Instruction>(use.get())) {
+        if (!m_canAssumeUniformDivergentUsesMap.count(operandInst))
+          continue;
+        if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block) ||
+            (loadInst && loadInst->getPointerAddressSpace() == ADDR_SPACE_BUFFER_FAT_POINTER)) {
+          canCollectOperands = false;
+          break;
+        }
+        operandInsts.push_back(operandInst);
+      }
+    }
```
IMHO, This loop can be merged with the above for-loop. I don't see any blocking reason. Maybe I missed something?

xuechen417 replied:
Thanks. You're right. The two for-loops can be merged into one. Done.

ruiling commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,418 @@
+//
+// @param block : The process basic block
+// @param readFirstLaneCount : The initial amdgcn_readfirstlane vector
+void PatchReadFirstLane::findInsertLocations(BasicBlock *block,
```
I would like to propose some implementation I came up which sounds more easy to understand.
btw. I think the function is better to be renamed as `findBestInsertLocation()`.
I just simply paste my implementation here, @nhaehnle @xuechen417 please help take a look whether it sounds good. I have not tested yet.

```
// We want instructions appear late in the IR put early in the set. That is to say the instructions are stored in the
// reverse order how they appear in the IR.
static bool instructionOrdering(const Instruction *p1, const Instruction *p2) {
  if (p1->comesBefore(p2))
    return false;
  return true;
}

void PatchReadFirstLane::findInsertLocations(BasicBlock *block,
                                             const SmallVectorImpl<Instruction *> &initialReadFirstLanes) {
  m_insertLocations.clear();

  // As we are trying to lift readfirstlane(i.e. RFL) instruction up in the IR. we must ensure we will not
  // introduce any extra RFLs. If one instruction access more than one divergent sources, we will do some lookahead
  // search here, to see whether these divergent sources come from the same value (i.e. they have a common predecessor
  // in def-use graph) so we can still move the RFL up in the IR to the common predecessor. If we cannot prove
  // that, just insert after the current instruction. This can ensure we are not inserting extra RFL during the lift
  // process.
  DenseSet<Instruction *> visited;
  for (auto &initialReadFirstLane : initialReadFirstLanes) {
    // Find an insert location for each initial readfirstlane
    std::queue<Instruction *> candidateQ;
    candidateQ.push(initialReadFirstLane);
    while (!candidateQ.empty()) {
      auto candidate = candidateQ.front();
      candidateQ.pop();

      if (visited.count(candidate))
        continue;

      visited.insert(candidate);

      auto &sources = m_canAssumeUniformDivergentUsesMap[candidate];
      unsigned divergentSrcNum = sources.size();
      if (divergentSrcNum == 0) {
        // The candidate has no divergent source, we cannot lift RFL further.
        addLocation(candidate);
      } else if (divergentSrcNum == 1) {
        candidateQ.push(sources[0]);
      } else {
#define MAX_SEARCH_STEP 10
        // we have more than one divergent sources, check whether we can prove they come from the same divergent value.
        std::set<Instruction *, decltype(&instructionOrdering)> divergentOrigins;
        for (auto s : sources)
          divergentOrigins.insert(s);

        unsigned searchStep = 0;
        while (divergentOrigins.size() > 1) {
          auto bottom = divergentOrigins.begin();

          auto &sources2 = m_canAssumeUniformDivergentUsesMap[*bottom];
          // TODO I didn't think further about this, stop for now for simplicity.
          if (!sources2.size())
            break;

          divergentOrigins.erase(bottom);

          for (auto s : sources2)
            divergentOrigins.insert(s);

          if (searchStep++ > MAX_SEARCH_STEP)
            break;
        }
        // If we successfully prove that the divergent sources of the current instruction shares the same origin,
        // Let's continue search, else let's stop at current candidate.
        if (divergentOrigins.size() == 1)
          candidateQ.push(*divergentOrigins.begin());
        else
          addLocation(candidate);
      }
    }
  }
}
```

xuechen417 replied:
I have re-work the logic of this function which is more general than the last version. Please review.

ruiling commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,418 @@
+  LegacyDivergenceAnalysis *m_divergenceAnalysis; // The divergence analysis
+  TargetTransformInfo *m_targetTransformInfo;     // The target tranform info to determine stop propagation
+  DenseMap<Instruction *, SmallVector<Instruction *, 2>>
+      m_canAssumeUniformDivergentUsesMap; // The map key is an instruction `I` that can be assumed uniform. The map
```
can we use a shorter name? maybe `m_uniformDstSrcMap` or anybody else has other suggestion?

ruiling replied:
sounds good to me.

xuechen417 reviewed (commented):
V9: Refine `collectAssumeUniforms`; rework `findbestInsertLocation` to generialize for multiple divergent operands; add more comments.

amdvlk-admin commented:
### Test summary for commit 6bc7b6dfef9cb454f276a16d230464739a4fc761
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 76544f9b1f743ab50e1cdd8db25acc6bf27d9e8b</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev fde3d4e2bd90743606ca7b724bf62a53cacf4615</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/96327)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26321/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21853/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24852/48153 (51.6%)</li>
<li>Failed: 0/48153 (0.0%)</li>
<li>Skipped: 23301/48153 (48.4%)</li>
</ul></details>


</ul></details>

xuechen417 commented:
V10: Fix lit test failures.

amdvlk-admin commented:
### Test summary for commit 882f9961399e3247f1ca810c1665b2d4ebb9da8a
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 76544f9b1f743ab50e1cdd8db25acc6bf27d9e8b</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev fde3d4e2bd90743606ca7b724bf62a53cacf4615</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/144501)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26321/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21853/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26320/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21854/48174 (45.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24852/48153 (51.6%)</li>
<li>Failed: 0/48153 (0.0%)</li>
<li>Skipped: 23301/48153 (48.4%)</li>
</ul></details>


</ul></details>

xuechen417 commented:
@dstutt , Could you help to review the latest change?

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,477 @@
+  LegacyDivergenceAnalysis *m_divergenceAnalysis; // The divergence analysis
+  TargetTransformInfo *m_targetTransformInfo;     // The target tranform info to determine stop propagation
+  DenseMap<Instruction *, SmallVector<Instruction *, 2>>
+      m_canAssumeUniformDivergentUsesMap; // The map key is an instruction `I` that can be assumed uniform. The map
+                                          // value is a vector of instructions that we can apply readfirstlane on them.
+                                          // An empty vector means the instruction `I` is an insert location.
```
Suggestion for improving the comment slightly (I don't know how to check the line length on GitHub):
```suggestion
      m_canAssumeUniformDivergentUsesMap; // The map key is an instruction `I` that can be assumed uniform.
                                          // That is, we can apply readfirstlane to the result of `I` and remain correct. If the map
                                          // value vector is non-empty, it contains a list of instructions that we can apply
                                          // readfirstlane on to achieve the same effect as a readfirstlane on `I`.
                                          // An empty vector means that it is not possible to life a readfirstlane beyond `I`.
```

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,477 @@
+      if (isAllUsersAssumedUniform(operandInst)) {
+        // The operand can be assumed uniform becasue all users are assumed uniform
+        m_canAssumeUniformDivergentUsesMap[operandInst] = {};
+        // NOTE: Stop propagation for a loadInst whose address space is fat pointer because it has uncertain size
+        if (loadInst && loadInst->getPointerAddressSpace() == ADDR_SPACE_BUFFER_FAT_POINTER)
+          canCollectOperands = false;
```
Why is this check inside the loop? It refers to inst, not operandInst, which makes sense to me but suggests that the check should the outside of the loop.

nhaehnle replied:
I think this check isn't really necessary. It will implicitly fall out from the fact that pointers aren't supported types.

ruiling replied:
The reason I asked for the check against fat-pointer is for case like we cast an <2 x i32> to pointer type, and then the pointer was cast to fat pointer, then some value derived from that fat pointer was assume to-be uniform (I am not sure whether this happens or not), I thought we should not back-propagate through the fat pointer to the <2 x i32>, but after rethinking the problem I think we can propagate through the fat-pointer for such case to <2 x i32> if no other blocking condition, am I right? As long as we do not insert readfirstlane directly on fat pointers(which is excluded by the type-checking right before inserting readfirstlane), we would still be safe.

xuechen417 replied:
Rebased code on Nicolai's two commits. I made a minor change to fix small bugs. 
-For candidate instruction itself, `call i32 @llvm.amdgcn.struct.tbuffer.load.i32(<4 x i32> %21, i32 %VertexIndex` should be a condition to stop propagate to its operands. Otherwise, the best location will be `%VertexIndex` which results in failures.
-In the middle loop of `findBestInsertLocation`, we should break out if the `current` cannot find in the map.

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,477 @@
+        if (loadInst && loadInst->getPointerAddressSpace() == ADDR_SPACE_BUFFER_FAT_POINTER)
+          canCollectOperands = false;
+      } else {
+        canCollectOperands = false;
```
You can `break` here.

xuechen417 replied:
The code does not exist any more.

nhaehnle commented on lgc/patch/PatchReadFirstLane.cpp:
```diff
@@ -0,0 +1,477 @@
+      applyReadFirstLane(inst, builder);
+      changed = true;
+    }
+  }
```
You can `m_canAssumeUniformDivergentUsesMap.clear()` at the end of each iteration, since each basic block is treated separately. It keeps the size of the structure smaller, and actually we **have** to clear it at the end anyway to avoid keeping dangling pointers.

nhaehnle reviewed (commented):
I have some inline comments, but I'm also looking at findBestInsertLocation. I like @ruiling's suggestion, but would adjust it slightly.

nhaehnle commented:
I will follow up soon.

nhaehnle commented:
Please take a look at https://github.com/nhaehnle/llpc/tree/dev-patchreadfirstlane. It contains a version of @ruiling's idea as one commit, and a change to collectAssumeUniforms as a second commit.

The change about collectAssumeUniforms is for a case like:
```
%a = ...
%b.0 = add %a, %other
%b.1 = add %a, 2
%c.0 = readfirstlane(%b.0)
%c.1 = readfirstlane(%b.1)
```
When evaluating `inst == %b.0`, the `operandInst == %a` is already in the map, and the existing collectAssumeUniforms would no longer collect it. This is a "dishonest" representation of the program and could lead (in more complex examples) to an incorrect conclusion by logic later on during findBestInsertLocation.

There are probably different ways to address this -- my change does it by only ever adding instructions to m_canAssumeUniformDivergentUsesMap when it is "their turn".

However, I haven't been able to test these changes yet, so there may well be simple bugs.

xuechen417 commented:
Thanks @nhaehnle and @ruiling for the review. I will update the PR based on Nicolai's commits and do a testing.

xuechen417 reviewed (commented):
V10: Rebase code based on 59462cb and 44e2b6f. Fix small bugs.

amdvlk-admin commented:
### Test summary for commit 26a00aa6f5af2d51d2aa6ff6ffc115c882a32586
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev eb28c7ef53329f838e693f233696823330e964c0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev fde3d4e2bd90743606ca7b724bf62a53cacf4615</li>
</ul></details>

<details><summary>CTS tests (Failed: 16/144501)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26316/48174 (54.6%)</li>
<li><details><summary>Failed: 5/48174 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo.dynamically_uniform_tess_eval
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo_storage_buffer_decoration.dynamically_uniform_tess_ctrl
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo_storage_buffer_decoration.dynamically_uniform_tess_eval
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ubo.dynamically_uniform_tess_ctrl
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ubo.dynamically_uniform_tess_eval
Stack trace: Crash
```
</p></details></li>

<li>Skipped: 21853/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26314/48174 (54.6%)</li>
<li><details><summary>Failed: 6/48174 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo.dynamically_uniform_tess_eval
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo_storage_buffer_decoration.dynamically_uniform_tess_ctrl
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo_storage_buffer_decoration.dynamically_uniform_tess_eval
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ubo.dynamically_uniform_tess_ctrl
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ubo.dynamically_uniform_tess_eval
Stack trace: Crash

FAILURE: dEQP-VK.glsl.texture_functions.textureprojoffset.sampler1d_vec4_fixed_fragment
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 21854/48174 (45.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24847/48153 (51.6%)</li>
<li><details><summary>Failed: 5/48153 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo.dynamically_uniform_tess_eval
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo_storage_buffer_decoration.dynamically_uniform_tess_ctrl
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ssbo_storage_buffer_decoration.dynamically_uniform_tess_eval
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ubo.dynamically_uniform_tess_ctrl
Stack trace: Crash

FAILURE: dEQP-VK.glsl.opaque_type_indexing.ubo.dynamically_uniform_tess_eval
Stack trace: Crash
```
</p></details></li>

<li>Skipped: 23301/48153 (48.4%)</li>
</ul></details>


</ul></details>

JaxLinAMD commented:
sorry for the delay of test result, I happened to maintain test system just now but the failure looks like a real failure.

xuechen417 commented:
V10: Rename to `m_uniformDivergentUsesMap`. Fix cts failures by breaking out when the `candidate` is not in the map.

amdvlk-admin commented:
### Test summary for commit 0f63eaf674564e807cd6698a5df4daf06cad5d1f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev eb28c7ef53329f838e693f233696823330e964c0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev fde3d4e2bd90743606ca7b724bf62a53cacf4615</li>
</ul></details>

<details><summary>CTS tests (Failed: 2/144501)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26321/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21853/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26318/48174 (54.6%)</li>
<li><details><summary>Failed: 2/48174 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.draw.shader_draw_parameters.base_instance.draw_indexed_indirect
Stack trace: Flake

FAILURE: dEQP-VK.tessellation.invariance.outer_edge_index_independence.triangles_fractional_even_spacing_ccw
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 21854/48174 (45.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24852/48153 (51.6%)</li>
<li>Failed: 0/48153 (0.0%)</li>
<li>Skipped: 23301/48153 (48.4%)</li>
</ul></details>


</ul></details>

xuechen417 commented:
> sorry for the delay of test result, I happened to maintain test system just now but the failure looks like a real failure.

Thanks Jax. I pushed a fix for the failing. The new two failures "Flake" and my change has no effect on them.

xuechen417 commented:
@nhaehnle, all tests are passing now after rebasing the code from your commits, do you have comment?

amdvlk-admin commented:
### Test summary for commit 0f63eaf674564e807cd6698a5df4daf06cad5d1f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev eb28c7ef53329f838e693f233696823330e964c0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev fde3d4e2bd90743606ca7b724bf62a53cacf4615</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/192504)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26321/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21853/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26320/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21854/48174 (45.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24852/48153 (51.6%)</li>
<li>Failed: 0/48153 (0.0%)</li>
<li>Skipped: 23301/48153 (48.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx103</summary><ul>

<li>Passed: 26219/48003 (54.6%)</li>
<li>Failed: 0/48003 (0.0%)</li>
<li>Skipped: 21784/48003 (45.4%)</li>
</ul></details>


</ul></details>

nhaehnle reviewed (commented):
Thank you for doing this. I don't have any other comments at this point :)
